lib/auth/predicate/builtins.go

```diff
@@ -0,0 +1,361 @@
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return nil, trace.BadParameter("cannot xor types: %T and %T", a, b)
+	}
+
+	switch aT := a.(type) {
+	case int:
+		return aT ^ b.(int), nil
+	case bool:
+		return aT != b.(bool), nil
+	default:
+		return nil, trace.BadParameter("xor unsupported for type type: %T", a)
+	}
+}
+
+func builtinSplit(a, b string) (any, error) {
+	return strings.Split(a, b), nil
+}
+
+func builtinUpper(a string) (any, error) {
+	return strings.ToUpper(a), nil
+}
+
+func builtinLower(a string) (any, error) {
+	return strings.ToLower(a), nil
+}
+
+func builtinContains(a any, b string) (any, error) {
+	switch aT := a.(type) {
+	case string:
+		return strings.Contains(aT, b), nil
+	case []string:
+		for _, s := range aT {
+			if s == b {
+				return true, nil
+			}
+		}
+
+		return false, nil
+	default:
+		return nil, trace.BadParameter("contains not valid for type: %T", a)
+	}
+}
+
+func builtinFirst(a []string) (any, error) {
+	if len(a) == 0 {
+		return nil, nil
+	}
+
+	return a[0], nil
+}
+
+func builtinAppend(a any, b string) (any, error) {
+	switch aT := a.(type) {
+	case []string:
+		return append(aT, b), nil
+	default:
+		return nil, trace.BadParameter("append not valid for type: %T", a)
+	}
+}
+
+func builtinArray(elements ...any) (any, error) {
+	arr := make([]string, len(elements))
+	for i, e := range elements {
+		s, ok := e.(string)
+		if !ok {
+			return nil, trace.BadParameter("cannot create array with element type %T", e)
+		}
+
+		arr[i] = s
+	}
+
+	return arr, nil
+}
+
+func builtinReplace(in any, match, with string) (any, error) {
+	switch inT := in.(type) {
+	case string:
+		return strings.Replace(inT, match, with, -1), nil
+	case []string:
+		for i, s := range inT {
+			if s == match {
+				inT[i] = with
+			}
+		}
+
+		return inT, nil
+	default:
+		return nil, trace.BadParameter("replace not valid for type: %T", in)
+	}
+}
+
+func builtinLen(a any) (any, error) {
+	switch aT := a.(type) {
+	case string:
+		return len(aT), nil
+	case []string:
+		return len(aT), nil
+	default:
+		return nil, trace.BadParameter("len not valid for type: %T", a)
+	}
+}
+
+func builtinRegex(a string) (any, error) {
+	return regexp.Compile(a)
+}
+
+func builtinMatches(to *regexp.Regexp, against string) any {
+	return to.MatchString(against)
+}
+
+func builtinContainsRegex(arr []string, regex *regexp.Regexp) (any, error) {
+	for _, s := range arr {
+		if regex.MatchString(s) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// type-generic for future extensibility
+func builtinMapInsert(m, k, v any) (any, error) {
+	mT, ok := m.(map[string]string)
+	if !ok {
+		return nil, trace.BadParameter("cannot insert into map of type: %T", m)
+	}
+
+	kS, ok := k.(string)
+	if !ok {
+		return nil, trace.BadParameter("cannot use non-string key of type: %T", k)
+	}
+
+	vS, ok := v.(string)
+	if !ok {
+		return nil, trace.BadParameter("cannot use non-string value of type: %T", k)
+	}
+
+	mT[kS] = vS
```

Comment:
when do we want people to modify the env? Are policies ordered in such a way that this effect can be relied upon within a single policy or across multiple policies?

Reply:
I've revised these functions to always copy their input map and return the modified map instead. This more closely matches how they are modeled in predicate and is side-effect free.